torch/_inductor/memory.py

```diff
@@ -0,0 +1,555 @@
+            0,
+            0,
+        )
+    for t, node in enumerate(nodes):
+        node_name_to_time[node.get_name()] = t
+        for sched_buf in node.get_outputs():
+            name_to_buf_info[sched_buf.get_name()] = BufferInfo(
+                sched_buf,
+                sched_buf.mpi_buffer.size_alloc,
+                sched_buf.mpi_buffer.size_free,
+                t,
+                t,
+            )
+
+    # assign end_time
+    for buf_name, buf_info in name_to_buf_info.items():
+        succ_node_time = [
+            node_name_to_time[succ_node.get_name()]
+            for succ_node in buf_info.buffer.mpi_buffer.succ_nodes
+            if succ_node.get_name() in node_name_to_time
+        ]
+        if succ_node_time:
+            buf_info.end_time = max(succ_node_time)
+
+    # the end time of output buffers should be at the end of the horizon
+    for buf_name in graph_outputs:
+        if buf_name in name_to_buf_info:
+            name_to_buf_info[buf_name].end_time = len(nodes) - 1
+
+    # incremental memory changes at each time period
+    memory = [0 for _ in range(len(nodes) + 1)]
+
+    # for each buffer, update memory when created and when freed
+    for buf_name, buf_info in name_to_buf_info.items():
+        memory[buf_info.start_time] += buf_info.size_alloc
+        memory[buf_info.end_time + 1] -= buf_info.size_free
+
+    # get peak memory by compute the cumulative memories
+    max_memory = 0
+    cur_memory = 0
+    memories_at_nodes = []
+    for t in range(len(nodes) + 1):
+        cur_memory += memory[t]
+        memories_at_nodes.append(cur_memory)
+        max_memory = max(max_memory, cur_memory)
+
+    return (max_memory, memories_at_nodes)
+
+
+def assign_predcessor_and_successor_nodes_to_nodes(
+    nodes: List[BaseSchedulerNode], name_to_fused_node: Dict[str, BaseSchedulerNode]
+) -> None:
+    """
+    Assign to each scheduler node its predecessor and successor nodes.
+    """
+    from .scheduler import SchedulerBuffer
+
+    for node in nodes:
+        node.mpi_node.pred_nodes = list(
+            {
+                name_to_fused_node[pred_buffer.defining_op.get_name()]
+                for pred_buffer in node.mpi_node.pred_buffers
+                if (
+                    isinstance(pred_buffer, SchedulerBuffer)
+                    and pred_buffer.defining_op.get_name() in name_to_fused_node
+                )
+            }
+        )
+        node.mpi_node.succ_nodes = list(
+            {
+                succ_node
+                for buffer in node.get_outputs()
+                for succ_node in buffer.mpi_buffer.succ_nodes
+            }
+        )
+
+
+def topological_sort_lpmf(
+    nodes: List[BaseSchedulerNode],
+    name_to_input_buf: Dict[str, FreeableInputBuffer],
+    name_to_buf: Dict[str, SchedulerBuffer],
+    graph_outputs: Set[str],
+) -> List[BaseSchedulerNode]:
+    """
+    A bfs-based greedy topological order. LPMF stands for "Least Peak Memory First".
```

Comment:
does this come from a paper ? if so, link ?

Reply:
Added.